### ISSUE TITLE
[Feature Request] npm-like version managment

### ISSUE THREAD

**Issue by ReimuNotMoe** [TheLartians/CPM.cmake#99]
Hello, firstly thanks for the awesome package manager you made. I had a great time using this in some of my projects.

But I think a more advanced versioning feature [like npm](https://docs.npmjs.com/files/package.json#dependencies) is essential for a package manager.

I think having this would be a great improvement for CPM. Would you like to have a look into it?

Thanks a lot in advance!

**Comment by TheLartians** [comment]
Hey thanks for your feedback! I agree that more advanced version management like npm would be awesome! However, being not having a package database means there is no way for CPM to be aware of package versions outside of the ones defined in the current project. This would however be required to provide ambiguous version ranges, for example.
Perhaps in a future version we could be able to parse an external package database, but for now I'm afraid this is out of the scope of this project.

**Comment by luketrevorrow** [comment]
How about if you took the MacOS brew utility as an example, and keep the package information in a github repo. That way you maintain links to the git repos used by CPM. I also think that CPM as a package manager is pretty cool, and set-up a link for it from the cppreference open source libraries page: [https://en.cppreference.com/w/cpp/links/libs](url)

**Comment by TheLartians** [comment]
> How about if you took the MacOS brew utility as an example, and keep the package information in a github repo.

Yeah, I was also thinking about something like this, though a git repo as a database has many issues itself, as it eventually becomes very inefficient. Another issue is there are many edge cases we have to think about. For example, how to handle mixing external and registered dependencies, how to deal with dependencies that define different databases etc. Also, implementing that kind of logic in CMake will be a pain.

For me the current implementation covers all my use cases so I probably won't be looking into this for a while. You're very welcome to make suggestions or contributions toward this though! 😉 

> I also think that CPM as a package manager is pretty cool, and set-up a link for it from the cppreference open source libraries page: https://en.cppreference.com/w/cpp/links/libs

That's really cool! Thanks for featuring CPM over there! :)

**Comment by ScottBailey** [comment]
Is this a suggestion for a stand alone package manager? Like vcpkg? I don't think that's what CPM is or necessarily should be.

If not, what would the syntax look like for `fmt` as opposed to the existing shorthand? 
```cmake
CPMAddPackage("gh:fmtlib/fmt#7.1.3")
```

**Comment by TheLartians** [comment]
So the way I understand the request would be for CPM to scrape all could scrape the source repository for all available versions and then decide on the highest one compatible with all version ranges specified in the project. The your example, this would allow users to specify something like `CPMAddPackage("gh:fmtlib/fmt@^7.1.0")` and CPM could automatically resolve it to `7.1.3`. 

Implementing this would be quite the undertaking though as without a central database of projects, finding an optimal configuration is non-trivial. Also we would need to implement version lock files first for reproducibility and safety from supply-chain-attacks.

**Comment by ScottBailey** [comment]
Hmm, I think there are ways to query GitHub for available releases using GitHub's CLI tool (`gh`). I don't really have time to look into that right now, but maybe I'll add some design ideas here if/when I do.

**Comment by JohelEGP** [comment]
See [REST API endpoints for releases and release assets](https://docs.github.com/en/rest/releases?apiVersion=2022-11-28), and [Get a tag](https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28#get-a-tag) or the answer to [How to see remote tags?](https://stackoverflow.com/a/25987962)
